[AutoDiff] [SILVerifier] Verify `autodiff_function` instruction

rxwei opened the pull request:
* Verify the JVP type and VJP type in an `autodiff_function` instruction.
* Correct the associated function type calculator utility to produce closures with `@callee_guaranteed` callee convention.
* Add example code (/test/AutoDiff/autodiff_function_inst.sil) that demonstrates how we emit a thunk that forms a VJP function.

rxwei committed:
Verify JVP and VJP type signatures in `autodiff_function`

rxwei committed:
Misc comment fixes

marcrasi commented on include/swift/AST/Types.h:
```diff
@@ -4077,8 +4077,8 @@ class SILFunctionType final : public TypeBase, public llvm::FoldingSetNode,
 
-  /// Returns the differentiation function type associated functions with this
-  /// function type.
+  /// Returns the type of a differentiation fucntion that is associated with
```
s/fucntion/function

marcrasi commented on lib/SIL/SILFunctionType.cpp:
```diff
@@ -290,7 +290,7 @@ CanSILFunctionType SILFunctionType::getAutoDiffAssociatedFunctionType(
     auto pullbackType = SILFunctionType::get(getGenericSignature(),
                                              ExtInfo(), SILCoroutineKind::None,
-                                             ParameterConvention::Direct_Owned,
+                                         ParameterConvention::Direct_Guaranteed,
```
this has become weirdly indented

rxwei replied:
This is to fit within 80 columns. It's a pretty common, but weird, pattern within this code base.

rxwei replied:
Re-formatted.

marcrasi commented on lib/SIL/SILInstructions.cpp:
```diff
@@ -649,6 +649,8 @@ std::pair<SILValue, SILValue> AutoDiffFunctionInst::
   assert(differentiationOrder > 0 &&
          differentiationOrder <= this->differentiationOrder);
+  assert(!getAssociatedFunctions().empty() &&
+         "Associated functions may not exist in raw SIL");
```
I don't understand this message. Does raw SIL have anything to do with this?

rxwei replied:
When associated functions are not specified, the differentiation pass will fill them in. When the stage is raw, we cannot assume AD has already run.

rxwei replied:
Clarified.

marcrasi commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -1255,13 +1255,34 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
+        require(!jvpType->isDifferentiable(),
+                "The JVP function must not be @autodiff");
+        auto expectedJVPType = origTy->getAutoDiffAssociatedFunctionType(
+            adfi->getParameterIndices(), adfi->getDifferentiationOrder(),
```
adfi->getDifferentiationOrder() should be changed to i?

marcrasi commented on lib/SIL/SILVerifier.cpp:
```diff
@@ -1255,13 +1255,34 @@ class SILVerifier : public SILVerifierBase<SILVerifier> {
+        require(!vjpType->isDifferentiable(),
+                "The VJP function must not be @autodiff");
+        auto expectedVJPType = origTy->getAutoDiffAssociatedFunctionType(
+            adfi->getParameterIndices(), adfi->getDifferentiationOrder(),
```
adfi->getDifferentiationOrder() should be changed to i?

rxwei committed:
Address review comments.

- Use the differentiation order induction variable in the verifier.

- Re-format.

- Fix a typo.

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux CPU

rxwei commented:
@swift-ci please test tensorflow Linux
